src/opus.py

```diff
@@ -229,3 +232,75 @@ async def load_nodes(nodes_dict: Dict[str, dict], script_path: Path) -> Tuple[Di
         nodes[key] = Node(**typed_node)
 
     return nodes, action_dicts
+
+
+def validate_references(opus: Opus):
+    """
+    Validate that we only refer to existing nodes, assets and actions.
+
+    Exits the program if validation fails.
+
+    Parameters
+    ----------
+    opus
+        The opus
+    """
+    # Nodes
+    referred_nodes = set([opus.start_node])
+    for node in opus.nodes.values():
+        if isinstance(node.next, str):
+            referred_nodes.add(node.next)
+        elif isinstance(node.next, list):
+            for choice in node.next:
+                referred_nodes.add(choice.node)
+    actual_nodes = set(opus.nodes.keys())
+    if referred_nodes != actual_nodes:
+        nonexistent_nodes = referred_nodes - actual_nodes
+        unreferred_nodes = actual_nodes - referred_nodes
+        print("Malformed opus!")
+        if nonexistent_nodes:
+            print("References to nonexistent nodes:")
+            print('\n'.join(nonexistent_nodes))
+        if unreferred_nodes:
+            print("Nodes never referred to:")
+            print('\n'.join(unreferred_nodes))
+        sys.exit(1)
+
+    # Assets
+    referred_assets = set(["script"])
+    for action in opus.action_templates.values():
+        referred_assets.update(set(action.assets))
+    actual_assets = set(opus.assets.keys())
+    if referred_assets != actual_assets:
+        nonexistent_assets = referred_assets - actual_assets
+        unreferred_assets = actual_assets - referred_assets
+        print("Malformed opus!")
+        if nonexistent_assets:
+            print("References to nonexistent assets:")
+            print('\n'.join(nonexistent_assets))
+        if unreferred_assets:
+            print("Assets never referred to:")
```

Comment:
During the creation of the opus it might be nice to have assets defined in advance but you still want to test that the actions you have added (this far) is OK. Could we split this up so you get an error for non-existant errors and a warning for unreferred ones? (perhaps same for nodes and actions?)

Reply:
Hmm, the same logic applies for non-existent nodes as well. When first creating an opus, you probably want to test it before it's complete.

What do you think about making exit/warning behavior configurable (with yet another environment variable)? Or maybe keep all as warnings, and trust anyone who modifies the opus to check the logs?

Reply:
Sounds like a good idea!

Reply:
Done! (configurable)